src/dataEntities/block.ts

```diff
@@ -50,3 +54,87 @@ export interface ResponderBlock extends IBlockStub {
 }
 
 export interface Block extends IBlockStub, Logs, Transactions, TransactionHashes {}
+
+export class BlockItemStore<TBlock extends IBlockStub> extends StartStopService {
+    private readonly subDb: LevelUp<EncodingDown<string, any>>;
+    constructor(db: LevelUp<EncodingDown<string, any>>) {
+        super("block-item-store");
+        this.subDb = sub(db, `block-item-store`, { valueEncoding: "json" });
+    }
+
+    private itemsByHeight: Map<number, Set<string>> = new Map();
+    private items: Map<string, any> = new Map();
+
+    protected async startInternal() {
+        // load all items from the db
+        for await (const record of this.subDb.createReadStream()) {
+            const { key, value } = (record as any) as { key: string; value: any };
+
+            const i = key.indexOf(":");
+            const height = Number.parseInt(key.substring(0, i));
+            const memKey = key.substring(i + 1);
+
+            const itemsAtHeight = this.itemsByHeight.get(height);
+            if (itemsAtHeight) itemsAtHeight.add(memKey);
+            else this.itemsByHeight.set(height, new Set([memKey]));
+            this.items.set(memKey, value);
+        }
+    }
+    protected async stopInternal() {}
+
+    public async putBlockItem(blockHeight: number, blockHash: string, itemKey: string, item: any) {
+        const memKey = `${blockHash}:${itemKey}`;
+        const dbKey = `${blockHeight}:${memKey}`;
+
+        const itemsAtHeight = this.itemsByHeight.get(blockHeight);
+        if (itemsAtHeight) itemsAtHeight.add(memKey);
+        else this.itemsByHeight.set(blockHeight, new Set([memKey]));
+        this.items.set(memKey, item);
+
+        await this.subDb.put(dbKey, item);
+    }
+
+    /**
+     * Gets the item with key `itemKey` for block `blockHash`.
+     * Returns `undefined` if a key is not present.
+     **/
+    public getItem(blockHash: string, itemKey: string) {
+        const key = `${blockHash}:${itemKey}`;
+        return this.items.get(key);
+    }
+
+    // This behavior is too specific for the needs of the BlockCache; should be refactored
+    // returns the items stored under the key "block" for the specific height, and wether they are
+    public getBlocksAtHeight(height: number): (TBlock & { attached: boolean })[] {
+        const itemsAtHeight = this.itemsByHeight.get(height) || new Set();
+        // collect blocks and attachment info
+
+        const blocks: (TBlock & { attached: boolean })[] = [];
+
+        for (const item of itemsAtHeight) {
+            // check if it is the actual block
+            if (item.endsWith(":block")) {
+                const itemKey = item.slice(0, -":block".length);
+                const block = this.items.get(item) as TBlock;
+                const attached = this.items.get(itemKey + ":attached") as boolean;
+                blocks.push({ ...block, attached: attached });
```

Comment:
As we discussed before, it's probably safer to return an object with a nested block property to avoid collisions

Reply:
Done in 41b88b5. Incidentally, it also makes other code more readable, as removing the "attached" field was a pain before and now it's not needed.